File: app/lib/models/viagem.dart
import 'package:controle_viagens/models/status_viagem.dart';

class Viagem {
  final int id;
  final DateTime dataIda;
  final DateTime dataVolta;
  final String destino;
  final StatusViagem status;
  final String finalidade;
  final String transporte;
  final String? observacoes;
  final DateTime criadoEm;

  Viagem({required this.id, required this.destino, required this.dataIda, required this.dataVolta, required this.status, required this.finalidade, required this.transporte, required this.observacoes, required this.criadoEm,});

  factory Viagem.fromJson(Map<String, dynamic> json) {
    return Viagem(
      id: json['id'],
      destino: json['destino'],
      dataIda: DateTime.parse(json['dataIda']),
      dataVolta: DateTime.parse(json['dataVolta']),
      status: StatusViagem.fromString(json['status']),
      finalidade: json['finalidade'],
      transporte: json['transporte'],
      observacoes: json['observacoes'],
      criadoEm:DateTime.parse(json['criadoEm']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'destino': destino,
      'status': status.apiValue,
      'dataIda': dataIda,
      'dataVolta': dataVolta,
      'finalidade': finalidade,
      'transporte': transporte,
      'observacoes': observacoes,
      'criadoEm': criadoEm,
    };
  }
}

File: backend/src/main/java/com/agt/viagens/domain/model/Viagem.java
package com.agt.viagens.domain.model;

import java.time.LocalDate;
import java.time.LocalDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(
    name = "viagens",
    indexes = {
        @Index(name = "idx_viagem_usuario_criado", columnList = "usuario_id, criado_em DESC")
    }
)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Viagem {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 200)
    private String destino;

    @Column(name = "data_ida", nullable = false)
    private LocalDate dataIda;

    @Column(name = "data_volta", nullable = false)
    private LocalDate dataVolta;

    @Column(nullable = false, length = 100)
    private String finalidade;

    @Column(nullable = false, length = 100)
    private String transporte;

    @Column(length = 1000)
    private String observacoes;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private StatusViagem status = StatusViagem.AGENDADA;

    @Column(name = "criado_em", nullable = false, updatable = false)
    private LocalDateTime criadoEm;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "usuario_id", nullable = false)
    private Usuario usuario;

    @PrePersist
    void prePersist() {
        if (this.criadoEm == null) this.criadoEm = LocalDateTime.now();
    }
}
